src/thor/bidirectional_astar.cc

```diff
@@ -917,129 +918,188 @@ void BidirectionalAStar::SetDestination(GraphReader& graphreader, const valhalla
+                        previous_transition_cost);
 
-  // Get the elapsed time at the end of the forward path. NOTE: PathInfo
-  // stores elapsed time as uint32_t but EdgeLabels uses float. Need to
-  // accumulate in float and cast to int so we do not accumulate roundoff
-  // error.
-  Cost cost = path.back().elapsed_cost;
-
-  // Get the transition cost at the last edge of the reverse path
-  Cost previous_transition_cost = edgelabels_reverse_[idx2].transition_cost();
-
-  // Append the reverse path from the destination - use opposing edges
-  // The first edge on the reverse path is the same as the last on the forward
-  // path, so get the predecessor.
-  uint32_t edgelabel_index = edgelabels_reverse_[idx2].predecessor();
-  while (edgelabel_index != kInvalidLabel) {
-    const BDEdgeLabel& edgelabel = edgelabels_reverse_[edgelabel_index];
-
-    // The first edge has no previous edge with elapsed cost so we can just use it directly
-    uint32_t predidx = edgelabel.predecessor();
-    if (predidx == kInvalidLabel) {
-      cost += edgelabel.cost();
-    } // This edge needs the elapsed cost between it and the previous edge, since its the reverse path
-    // in flipping the path around we also need to switch out the transition costs (shifting right)
-    else {
-      cost += edgelabel.cost() - edgelabels_reverse_[predidx].cost() - edgelabel.transition_cost();
-    }
-    cost += previous_transition_cost;
-    path.emplace_back(edgelabel.mode(), cost, edgelabel.opp_edgeid(), 0, edgelabel.restriction_idx(),
-                      previous_transition_cost);
+      // Check if this is a ferry
+      if (edgelabel.use() == Use::kFerry) {
+        has_ferry_ = true;
+      }
 
-    // Check if this is a ferry
-    if (edgelabel.use() == Use::kFerry) {
-      has_ferry_ = true;
+      // Update edgelabel_index and transition cost to apply at next iteration
+      edgelabel_index = predidx;
+      // We apply the turn cost at the beginning of the edge, as is done in the forward path
+      // Semantically this can be thought of is, how much time did it take to turn onto this edge
+      // To do this we need to carry the cost forward to the next edge in the path so we cache it here
+      previous_transition_cost = edgelabel.transition_cost();
     }
 
-    // Update edgelabel_index and transition cost to apply at next iteration
-    edgelabel_index = predidx;
-    // We apply the turn cost at the beginning of the edge, as is done in the forward path
-    // Semantically this can be thought of is, how much time did it take to turn onto this edge
-    // To do this we need to carry the cost forward to the next edge in the path so we cache it here
-    previous_transition_cost = edgelabel.transition_cost();
+    // For the first path just add it for subsequent paths only add if it passes viability tests
+    if (paths.empty() ||
+        (validate_alternate_by_sharing(graphreader, shared_edgeids, paths, path, max_sharing) &&
```

Comment:
this is where we validate the second criterion by ignoring paths that are too similar to already found ones